components/lib/jexl/src/main/java/mozilla/components/lib/jexl/lexer/LexerInput.kt

```diff
@@ -0,0 +1,63 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.jexl.lexer
+
+/**
+ * Helper class for reading a string character by character with the ability to "peek" at upcoming characters.
+ */
+@Suppress("TooManyFunctions")
+internal class LexerInput(private val value: String) {
+    private var position: Int = 0
+    private var mark: Int = 0
+
+    fun mark() {
+        mark = position
+    }
+
+    fun emit(): String {
+        return value.substring(mark, position)
+    }
+
+    fun proceed(steps: Int = 1) {
+        position += steps
+    }
+
+    fun peekEquals(candidate: String): Boolean {
+        if (position + candidate.length > value.length) {
+            return false
+        }
+
+        for (i in 0 until candidate.length) {
+            if (candidate[i] != value[position + i]) {
+                return false
+            }
+        }
+
+        position += candidate.length
+
+        return true
+    }
+
+    fun peekRange(steps: Int): String {
+        if (position + steps > value.length) {
+            return ""
+        }
+
+        return value.substring(position, position + steps)
+    }
+
+    fun character(): Char = value[position]
+
+    fun end() = position == value.length
+
+    fun next(): Char = if (last()) ' ' else value[position + 1]
```

Comment:
`next` and `last` seem like good candidates to increase code coverage (easy to test).